### DISCUSSION TITLE
How to add proselint, alex and write-good?

### DISCUSSION THREAD

**Discussion by kapral18** [LazyVim/LazyVim#1265] [Q&A] [answered]
I am following this [video](https://youtu.be/a_ZpTPaSn38?t=517) on setting up text linting.

But for some reason I keep getting a warning 

```sh
WARN [null-ls] failed to run generator: ...y/null-ls.nvim/lua/null-ls/helpers/generator_factory.lua:64: failed to decode json: Expected value but found invalid token at character 1│
```

This happens when I open nvim after adding this config

```lua
-- filePath: .config/nvim/lua/plugins/writing.lua
return {
{
"williamboman/mason.nvim",
opts = function(_, opts)
table.insert(opts.ensure_installed, "proselint")
table.insert(opts.ensure_installed, "write-good")
end,
},
{
"jose-elias-alvarez/null-ls.nvim",
  opts = function(_, opts)
    local nls = require("null-ls")
    opts.sources = vim.list_extend(opts.sources, {
      nls.builtins.diagnostics.proselint,
      nls.builtins.code_actions.proselint,
      nls.builtins.diagnostics.write_good,
    })
  end,
},
}
```

Has anyone come across this? And what was the solution?

**Comment by dpetka2001** [comment]
It should be `function(_, opts)`. You should also uncomment `-- table.insert(opts.ensure_installed, "write-good")` in `mason` spec, because the error I got stated no binary `write-good` could be found.

**Reply by kapral18** [reply]
my bad, I just noticed that I posted a wrong example.

Actually the example I used was exactly like your corrected answer:

```lua
return {
{
"williamboman/mason.nvim",
opts = function(_, opts)
table.insert(opts.ensure_installed, "proselint")
table.insert(opts.ensure_installed, "write-good")
end,
},
{
"jose-elias-alvarez/null-ls.nvim",
  opts = function(_, opts)
    local nls = require("null-ls")
    opts.sources = vim.list_extend(opts.sources, {
      nls.builtins.diagnostics.proselint,
      nls.builtins.code_actions.proselint,
      nls.builtins.diagnostics.write_good,
    })
  end,
},
}
```

but I still get

![CleanShot 2023-08-01 at 18 32 28@2x](https://github.com/LazyVim/LazyVim/assets/1625373/71662f35-e3f0-4734-9313-97147ac164d9)

**Reply by dpetka2001** [reply, marked as answer]
Can you give an example of code where you test this? I don't get such error in simple Markdown file.

**Reply by kapral18** [reply]
oh wow, you are spot on I actually see this happening in only this README.md. Thanks much for the cues. I can now debug and isolate pieces of the .md file that can cause this. I will post here once I find something useful. But in essence now I know that the config is actually correct. Thanks

**Reply by kapral18** [reply]
ok interesting stuff. I just noticed that this warning only appears when I open the .md file as the first thing when I open nvim from telescope find-files. If I open it after I opened some other file in find-files, or if I open it directly or via explorer no warnings happen.

**Reply by dpetka2001** [reply]
I cannot reproduce this. I also open the README.md from telescope find-files, but don't see any such error. Do you use default Lazyvim configuration? Do you happen to lazy-load any plugins?

**Reply by kapral18** [reply]
added my setup here https://github.com/kapral18/config.nvim

**Reply by kapral18** [reply]
Try opening some README.md file with a lot of code blocks maybe

**Reply by dpetka2001** [reply]
I don't have such README. Can you provide one in your repo as example?

**Reply by kapral18** [reply]
yes let me try to anonimize mine, hold on...

**Reply by dpetka2001** [reply]
I tried with your configuration with the following contents in a README.md
```markdown
# 💤 LazyVim

A starter template for [LazyVim](https://github.com/LazyVim/LazyVim).
Refer to the [documentation](https://lazyvim.github.io/installation) to get started.

```python
print("Hello, world!")
for i in range(10):
    print(i)


```json
{ "some": "json" }


```json
{
    "one": 2,
    "three": {
        "point_1": "point_2",
        "point_3": 3.4
    },
    "list": [
        "one",
        "two",
        "three"
    ]
}
```
I skipped the closing triple ticks at the end of each code block here, so that it could be formatted as Markdown as a whole. Otherwise it would be seperated into different codeblocks here as well. In the local README the closing triple ticks at the end of each code block exist as they should. I don't get any error with your configuration.

**Reply by kapral18** [reply]
Here you go. I added backslashes to backticks to be able to render it here in one block

```md
# Foo

Foo is foo

## Bar

provides the following abilities to:

- [x] [foo](#foo)
- [x] [bar](#bar)
- [x] [baz](#baz)

## How to use

Basic variables:

- BAR

  !!! Deprecated (use BAAR instead)

  Bla bla (optional, defaults to bla)

- BAZ

  !!! Deprecated (use BAAZ instead)

  Bla bla (required)


#### JSON

Bla bla about json

\```json
{
  "a": {
    "b": [
      {
        "c": "d",
        "e": "f"
      }
    ]
  }
}
\```

`a.b` is an array of objects with two properties:

- `c` - blabla
- `e` - blabla

Both `c` and `e` are required. And both can blabla
${bla} bla bla

\```json
{
  "a": {
    "b": [
      {
        "c": "$d",
        "e": "$f"
      }
    ]
  }
}
\```

blabla:

- `/a/b` -> `zetrd://c/d/`

\```javascript
const foobar = () => true
\```

### Foobarbaz

Blabla

Bar: `friggo.bar/bar`
Baz: `friggo.baz/baz`

Blabla.

For example:

\```Dockerfile
FROM somewhere

ENV SOMETHING=/SOMETHING

RUN something

EXPOSE 54002
\```

```

**Reply by dpetka2001** [reply]
The only time I managed to invoke the warning, was when I removed the backslashes in front of the triple ticks in json code blocks and immediately tried to save before the `write-good` progress was finished. So I guess it might be some kind of race condition between null-ls trying to format and write-good not being finished yet. But I couldn't reproduce it consistently. But it's just a warning, so I guess in the end if you wait for the write-good progress to finish and then save, it shouldn't matter as it would format the text correctly. All these, of course, just a guesstimation based on observations of the README's behavior.

**Reply by kapral18** [reply]
yes I guessed that it's a race cond as well, but in my case it always happens when I open the file not when I save

**Reply by dpetka2001** [reply]
Never happened to me when I opened the file. And I used your configuration from your repo to test that.

**Reply by kapral18** [reply]
actually I just tried to open with find-files but without NeoTree and surprise surprise - no warning

**Reply by kapral18** [reply]
now I am assuming that most of my problems that I see with lazyvim are probably stemming from neotree not playing well with other parts of lazyvim config

**Reply by kapral18** [reply]
actually I managed to also see the race you mentioned when I tried to mini.surround and undo quickly.

**Reply by dpetka2001** [reply]
Unfortunately, I have no further insight as to what might be causing it. You should open an issue maybe at `write-good` repo and hope for a better answer?

**Reply by dpetka2001** [reply]
Try to disable `write-good` and keep only `proselint` to see if there are any differences. Might help you narrow down the real culprit.

**Reply by kapral18** [reply]
happens with `proselint` as well unfortunately

**Reply by kapral18** [reply]
thanks for your time, I appreciate it
